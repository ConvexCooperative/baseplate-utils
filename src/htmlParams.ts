import { mergeWith } from "lodash-es";
import { RecursivePartial } from "./utils.js";

const defaultParams: HTMLTemplateParams = {
  lang: "en",
  pageTitle: "Web App",
  importMap: {
    type: "systemjs",
    name: "systemjs",
    useOverrides: true,
    entryModule: "root-config",
  },
  preloads: [],
  headScripts: [],
  headStylesheets: [],
};

export function mergeDefaultHtmlParams(
  htmlParams: RecursivePartial<HTMLTemplateParams>
): HTMLTemplateParams {
  const finalSettings: HTMLTemplateParams = mergeWith(
    {},
    defaultParams,
    htmlParams,
    (newValue, oldValue) => {
      return oldValue === null ? newValue : undefined;
    }
  );

  return finalSettings;
}

export interface HTMLTemplateParams {
  lang: string;
  pageTitle: string;
  importMap: {
    type: "systemjs" | "native";
    name: string;
    useOverrides: boolean;
    entryModule: string;
  };
  preloads: HTMLPreload[];
  headScripts: HTMLScript[];
  headStylesheets: HTMLStylesheet[];
}

export interface HTMLPreload {
  importSpecifier?: string;
  href?: string;
  as: string;
}

export interface HTMLScript {
  src: string;
  async: boolean;
  defer: boolean;
}

export interface HTMLStylesheet {
  href: string;
}
