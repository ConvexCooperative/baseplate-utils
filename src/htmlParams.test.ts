import { HTMLTemplateParams, mergeDefaultHtmlParams } from "./htmlParams.js";
import { RecursivePartial } from "./utils.js";

describe(`htmlParams`, () => {
  it(`matches snapshot for default params`, () => {
    const params = mergeDefaultHtmlParams({});
    expect(params).toMatchSnapshot();
  });

  it(`merges top level properties`, () => {
    const params: RecursivePartial<HTMLTemplateParams> = {
      lang: "es",
      pageTitle: "Aplicacion de Web",
      preloads: [
        {
          importSpecifier: "root-config",
          as: "script",
        },
      ],
    };
    const finalParams = mergeDefaultHtmlParams(params);
    expect(finalParams.lang).toBe(params.lang);
    expect(finalParams.pageTitle).toBe(params.pageTitle);
    expect(finalParams.preloads).toEqual(params.preloads);
  });

  it(`merges import map properties`, () => {
    const params: RecursivePartial<HTMLTemplateParams> = {
      importMap: {
        name: "main",
        useOverrides: false,
      },
    };
    const finalParams = mergeDefaultHtmlParams(params);
    expect(finalParams.importMap.type).toBe("systemjs");
    expect(finalParams.importMap.name).toBe(params.importMap!.name);
    expect(finalParams.importMap.useOverrides).toBe(
      params.importMap!.useOverrides
    );
  });
});
