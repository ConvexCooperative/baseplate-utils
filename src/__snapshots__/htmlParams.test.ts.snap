// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`htmlParams matches snapshot for default params 1`] = `
Object {
  "headScripts": Array [],
  "headStylesheets": Array [],
  "importMap": Object {
    "entryModule": "root-config",
    "name": "systemjs",
    "type": "systemjs",
    "useOverrides": true,
  },
  "lang": "en",
  "pageTitle": "Web App",
  "preloads": Array [],
}
`;
